### ISSUE TITLE
[question] Any plans to support iOS

### ISSUE THREAD

**Issue by kirk86** [Genymobile/scrcpy#1691]
- [X] I have checked that a similar [feature request](https://github.com/Genymobile/scrcpy/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22) does not already exist.

I know this kind of has been asked before but I thought maybe something might have changed during that time and now. Sorry if this annoying to you, but, I was wondering if you are aware of any other FOSS alternatives that are similar to `scrcpy` but also support iOS devices, or, are there any plans to support iOS devices? This is something much need in having a FOSS alternative for iOS devices.

**Comment by kirk86** [comment]
@alexandre-janniaux Thanks for the reference. Although I have no idea how things work behind the scenes but generally I do agree with 

> integration of this work in a tool

there are generally many disparate FOSS projects that would make sense to be integrated in a unifying tool.

For instance, [free-sidecar](https://github.com/ben-z/free-sidecar) enables sidecar for older devices which blacklisted from apple by default.

It would certainly be interesting to have them under a unifying tool which eventually allows to extend or mirror the iOS device to your mac/laptop.

**Comment by liangwu** [comment]
Wormhole (https://er.run) did the iOS work. cast and control iPhone from PC or Mac.

**Comment by kfatehi** [comment]
@liangwu I tried wormhole in response to your note. Screen mirror worked but it failed to provide me with control via its bluetooth hack... might just be my old PC's bluetooth hardware version is too old.

I am also investigating jailbreak universe and found this https://github.com/coolstar/RecordMyScreen which seems to have really good code that seems apropo for a scrcpy port core, to my naive eyes anyway.

**Comment by liangwu** [comment]
> @liangwu I tried wormhole in response to your note. Screen mirror worked but it failed to provide me with control via its bluetooth hack... might just be my old PC's bluetooth hardware version is too old.
> 
> I am also investigating jailbreak universe and found this https://github.com/coolstar/RecordMyScreen which seems to have really good code that seems apropo for a scrcpy port core, to my naive eyes anyway.

YES, It Works when your bluetooth adapter Supports Bluetooth Low Energy Peripheral Role. [Here](https://www.howto-connect.com/see-if-windows-10-pc-supports-bluetooth-low-energy-peripheral-role/) to check if your Windows PC supports or not.

**Comment by kirk86** [comment]
@liangwu 

> Wormhole (https://er.run) did the iOS work. cast and control iPhone from PC or Mac.

This didn't work for me, I tried to mirror screen from ipad 2 2011 to my late 2015 macbook but I got nothing no video/audio nothing got mirrored.

**Comment by liangwu** [comment]
> @liangwu
> 
> > Wormhole (https://er.run) did the iOS work. cast and control iPhone from PC or Mac.
> 
> This didn't work for me, I tried to mirror screen from ipad 2 2011 to my late 2015 macbook but I got nothing no video/audio nothing got mirrored.

Try wireless connection, keep your Mac and iPad in same network.

**Comment by kirk86** [comment]
I did, but that didn't work either, I believe it might have to do with the fact that the ipad is old, 2011.

**Comment by guanzhangrtk** [comment]
> I did, but that didn't work either, I believe it might have to do with the fact that the ipad is old, 2011.

Issue could be your macOS instead of iPad, this is from their FAQ
![image](https://user-images.githubusercontent.com/44019025/114273630-66a23d80-9a4d-11eb-99ae-8fd3b4684a5f.png)

**Comment by kirk86** [comment]
I think I have tried all of that but still nothing, I'll give it another go but my hopes are really low.
BTW I'm running Big Sur version 11.0.1.

**Comment by guanzhangrtk** [comment]
> I think I have tried all of that but still nothing, I'll give it another go but my hopes are really low.
> BTW I'm running Big Sur version 11.0.1.

We're probably straying off-topic here but if you need support for Wormhole you can look for their Discord. I'll just leave it at that, good luck!

**Comment by liangwu** [comment]
> @liangwu I tried wormhole in response to your note. Screen mirror worked but it failed to provide me with control via its bluetooth hack... might just be my old PC's bluetooth hardware version is too old.
> 
> I am also investigating jailbreak universe and found this https://github.com/coolstar/RecordMyScreen which seems to have really good code that seems apropo for a scrcpy port core, to my naive eyes anyway.

if you have jailbreaked your iOS devices, try Blackhole App(https://er.run/blackhole)

**Comment by liangwu** [comment]
> I think I have tried all of that but still nothing, I'll give it another go but my hopes are really low.
> BTW I'm running Big Sur version 11.0.1.

If you still would to try again, you could join their feedback telegram(https://t.me/joinchat/
NXEoQRtzLKZKTD9X2XFtRQ) and find it out

**Comment by hhhaiai** [comment]
现在有同学跟踪这个事吗？

**Comment by krishtoautomate** [comment]
any one point me to piece of code to control ios devices with bluetooth with touch-assist enabled.

both vysor and wormhole are paid. better to look for open-source

**Comment by guanzhangrtk** [comment]
@krishtoautomate You can try [KeyPad](https://apps.apple.com/us/app/keypad-keyboard-and-mouse/id1491684442), it's pretty basic but should do the job, but it's only available for macOS.

Alternatively, if you can work without the Bluetooth requirement, you can try Airtest IDE or Appium. Those are used for iOS device automation but performance wise they are pretty slow.

**Comment by wsvn53** [comment]
I've just released the first version of Scrcpy.app for iOS, which is based on scrcpy v1.8, you dont need to jailbroken your iOS device. But you may need to build or re-codesign to install on your iPhone now, because the AppStore version is under reviewing.

For more detail you can visit my opensource project: https://github.com/wsvn53/scrcpy-ios

**Comment by krishtoautomate** [comment]
@guanzhangrtk  KeyPad is good but need to expose apis to integrate with scrcpy

**Comment by guanzhangrtk** [comment]
@krishtoautomate Yeah I've asked the author for some APIs, I guess he's still working on it. Perhaps you can also do the same :)

**Comment by titopalito** [comment]
> Wormhole (https://er.run) did the iOS work. cast and control iPhone from PC or Mac.

Nice!
Without JailBreak you hava a screen mirror, JailBreaking the iPhone you get control from the mac (did't tried in Win PC)

**Comment by krishtoautomate** [comment]
Screen mirror can be done my integrating ios-minicap which is in c++ and can be easily integrated to scrcpy

**Comment by guanzhangrtk** [comment]
@fleytman Just use Airtest IDE if you are looking for control. It works perfectly fine.

**Comment by fleytman** [comment]
@guanzhangrtk It works as other appium webdriver agent solutions, but need run ide.

**Comment by truebit** [comment]
https://github.com/danielpaulus/quicktime_video_hack
Here is the iOS video hack for screening cast feature, I think scrcpy could implement the cast feature first😄

**Comment by adibaulia** [comment]
> I was also looking for scrcpy analogues I found an old project: https://github.com/testingbot/ios-vnc . Now there are problems with its building (but solvable) and it depends on the old web driver. But it seems to have worked once upon a time and its code might be useful There is also a way to output the screen with ios using idb: https://fbidb.io/docs/video
> 
> From what I tried and it worked: jailbreak + blackhole, but it's paid, plus it requires manual action (although you can automate it) jailbreak + https://github.com/julioverne/screendump, but the transmission quality leaves much to be desired
> 
> UPD. I haven't tried it yet, but maybe that's what we're looking for https://github.com/nanoscopic/controlfloor https://github.com/nanoscopic/ios_remote_provider
> 
> UPD2. All the projects, that I found:
> 
> * https://github.com/facebook/idb `idb video-stream --fps 30 --format h264 --compression-quality 1.0 --udid <udid> | ffmpeg -f h264 -i pipe:0 -vcodec copy -tune zerolatency -b 900k -framerate 30 -f mpegts udp://127.0.0.1:12345 ` only screen display
> * https://github.com/danielpaulus/quicktime_video_hack - view ala quiktime, but without it. Used in some implementations
> * Needs jaibreak
>   
>   * https://github.com/julioverne/screendump
>   * https://github.com/f-viktor/ghetto-scrcpy-iOS
>   * https://er.run/blackhole - paid, but it works
> * Needs webdriver agent
>   
>   * https://github.com/testingbot/ios-vnc is outdated, needs a lot of tweaks and depends on old repositories. But in theory can be improved
>     
>     * https://github.com/Karazum/ios-vnc is a fork with linux support
>   * controlfloor
>     
>     * https://github.com/nanoscopic/ios_remote_provider
>       The provider connects iOS devices to ControlFloor. This sets up video streaming from iOS devices to the browser, and allows you to control devices remotely.
>     * https://github.com/nanoscopic/controlfloor Remote device control system
>   * https://github.com/AirtestProject/iOS-Tagent works in Airtest ide
>   * https://github.com/DeviceFarmer/stf_ios_support is a cool project, hard to set up
>   * https://github.com/kavichki-test/ios-farm - similar to stf but unsupported project, as bonus [article](https://habr.com/ru/post/572668/) and [chat](https://t.me/ios_farm_kavichki)
>   * https://github.com/NetrisTV/ws-scrcpy web device control

any update for newer ios 16.3.1 and macOS Ventura? i have been trying these method but nothing helps

**Comment by fatahillahardhi** [comment]
> > I was also looking for scrcpy analogues I found an old project: https://github.com/testingbot/ios-vnc . Now there are problems with its building (but solvable) and it depends on the old web driver. But it seems to have worked once upon a time and its code might be useful There is also a way to output the screen with ios using idb: https://fbidb.io/docs/video
> > From what I tried and it worked: jailbreak + blackhole, but it's paid, plus it requires manual action (although you can automate it) jailbreak + https://github.com/julioverne/screendump, but the transmission quality leaves much to be desired
> > UPD. I haven't tried it yet, but maybe that's what we're looking for https://github.com/nanoscopic/controlfloor https://github.com/nanoscopic/ios_remote_provider
> > UPD2. All the projects, that I found:
> > 
> > * https://github.com/facebook/idb `idb video-stream --fps 30 --format h264 --compression-quality 1.0 --udid <udid> | ffmpeg -f h264 -i pipe:0 -vcodec copy -tune zerolatency -b 900k -framerate 30 -f mpegts udp://127.0.0.1:12345 ` only screen display
> > * https://github.com/danielpaulus/quicktime_video_hack - view ala quiktime, but without it. Used in some implementations
> > * Needs jaibreak
> >   
> >   * https://github.com/julioverne/screendump
> >   * https://github.com/f-viktor/ghetto-scrcpy-iOS
> >   * https://er.run/blackhole - paid, but it works
> > * Needs webdriver agent
> >   
> >   * https://github.com/testingbot/ios-vnc is outdated, needs a lot of tweaks and depends on old repositories. But in theory can be improved
> >     
> >     * https://github.com/Karazum/ios-vnc is a fork with linux support
> >   * controlfloor
> >     
> >     * https://github.com/nanoscopic/ios_remote_provider
> >       The provider connects iOS devices to ControlFloor. This sets up video streaming from iOS devices to the browser, and allows you to control devices remotely.
> >     * https://github.com/nanoscopic/controlfloor Remote device control system
> >   * https://github.com/AirtestProject/iOS-Tagent works in Airtest ide
> >   * https://github.com/DeviceFarmer/stf_ios_support is a cool project, hard to set up
> >   * https://github.com/kavichki-test/ios-farm - similar to stf but unsupported project, as bonus [article](https://habr.com/ru/post/572668/) and [chat](https://t.me/ios_farm_kavichki)
> >   * https://github.com/NetrisTV/ws-scrcpy web device control
> 
> any update for newer ios 16.3.1 and macOS Ventura? i have been trying these method but nothing helps

yep me too guys, any update? mostly on the other's app's or tools only can screen sharing not fully control the devices

**Comment by plops** [comment]
I used gemini 1.5 pro to summarize all solutions discussed in this thread. Here is the table summarizing the mentioned iOS screen mirroring/control solutions:

| Solution Name                    | Source Availability | View Only | Control | Needs Jailbreak | Supported Platforms | Notes                                               |
|------------------------------------|----------------------|-----------|---------|-----------------|--------------------|----------------------------------------------------|
| scrcpy                           | Open Source         | ✅         | ✅        | No              | Android           | Unlikely to support iOS officially                 |
| Wormhole (er.run)                | Commercial          | ✅         | ⚠️ Limited | No              | macOS, Windows, iOS | Wireless, screen mirroring + some control features |
| Blackhole (er.run)               | Commercial          | ✅         | ✅        | Yes             | macOS, iOS        | Good functionality, requires jailbreak           |
| quicktime_video_hack             | Open Source         | ✅         | ❌        | No              | macOS            | Leveraging QuickTime's built-in capabilities       |
| ios-minicap                      | Open Source         | ✅         | ❌        | No              | Various          | Potential for integration with scrcpy            |
| RecordMyScreen (coolstar)         | Open Source         | ✅         | ?        | Yes             | -                | Jailbreak required, code potentially useful for ports |
| free-sidecar                     | Open Source         | ✅         | ❌        | No              | macOS            | Enables Sidecar on older, unsupported Macs     |
| Scrcpy.app (wsvn53)             | Open Source         | ✅         | ?        | No              | iOS              | Under development, unclear if control is implemented |
| KeyPad                           | Commercial          | ❌        | ✅        | No              | macOS            | Bluetooth keyboard/mouse input, no screen mirroring |
| Airtest IDE                      | Open Source         | ✅         | ✅        | No              | Cross-platform   | Primarily for automation, might feel slow          |
| Appium                           | Open Source         | ✅         | ✅        | No              | Cross-platform   | Primarily for automation, might feel slow          |
| ios-vnc (testingbot)             | Open Source         | ✅         | ?        | Yes             | Various          | Outdated, requires significant tweaking           |
| ios-vnc (Karazum fork)           | Open Source         | ✅         | ?        | Yes             | Linux            | Fork of testingbot's ios-vnc                      |
| screendump (julioverne)         | Open Source         | ✅         | ❌        | Yes             | -                | Jailbreak required, subpar transmission quality     |
| ghetto-scrcpy-iOS (f-viktor)     | Open Source         | ✅         | ?        | Yes             | -                | Jailbreak required, unclear if control is implemented |
| controlfloor & ios_remote_provider| Open Source         | ✅         | ✅        | No              | Browser-based    | Connects iOS devices for remote control & streaming|
| iOS-Tagent (AirtestProject)    | Open Source         | ✅         | ✅        | No              | -                | Works with Airtest IDE                            |
| stf_ios_support (DeviceFarmer) | Open Source         | ✅         | ✅        | -                | -                | Complex setup, good for device farms              |
| ios-farm (kavichki-test)         | Open Source         | ✅         | ✅        | -                | -                | Similar to stf_ios_support, unsupported        |
| ws-scrcpy (NetrisTV)           | Open Source         | ✅         | ✅        | -                | Browser-based    | Web-based device control                           | 
| idb (facebook)                  | Open Source         | ✅         | ❌        | No              | Command line     | Requires command-line usage and ffmpeg for viewing |

**Comment by plops** [comment]
@bspinner, no. I just identified quicktime_video_hack as the most promising approach for what I want to do.

I did use LLM's for many such summaries before and I typically trust the results. I have never seen hallucinations when I ask summaries Gemini.

**Comment by fleytman** [comment]
iOS 18 will feature [screen mirroring](https://appleinsider.com/inside/macos-sequoia/tips/how-to-get-started-with-iphone-mirroring-in-ios-18-and-macos-sequoia) with control.

**Comment by mailinglists35** [comment]
> I've just released the first version of Scrcpy.app for iOS, which is based on scrcpy v1.8, you dont need to jailbroken your iOS device. But you may need to build or re-codesign to install on your iPhone now, because the AppStore version is under reviewing.
> 
> 
> 
> For more detail you can visit my opensource project: https://github.com/wsvn53/scrcpy-ios

this one works nicely on iOS 17 and Android TV 9, thank you!

**Comment by 0xGaryWu** [comment]
Did you toggle option on wormhole settings?

q8f13 ***@***.***> 于2025年2月21日周五 16:05写道：

**Comment by tbertels** [comment]
See also [RPiPlay](https://github.com/FD-/RPiPlay) (not tested).
